### ISSUE TITLE
fix machine-subject reference cycles

### ISSUE THREAD

**Issue by narfdotpl** [macoscope/SwiftyStateMachine#17]
Using a state machine as a property and providing oneself as a subject creates a reference cycle. :sob:  Example from the README using `[unowned self]` is incorrect.

My first idea is to just make the subject weak and don't perform any transitions if it's `nil`. That would spare users the is-the-subject-nil dance in transition closures. On the other hand, it would mean no support for value type subjects, I guess. But do value type subjects even make sense?

The other idea, the worse one, is to provide some wrapper class for creating weak references and to instruct users to use it, e.g.:

``` swift
final class Weak<T: AnyObject> {
    private(set) weak var value: T?

    init(_ value: T) {
        self.value = value
    }
}
```

What do you think, @bartekchlebek, @ncreated, @ochococo?

**Comment by bubski** [comment]
@narfdotpl #20 is all I came up with so far.

**Comment by narfdotpl** [comment]
Clever! I started working on a fix where I explicitly made subject weak, but it resulted in this whole `NoSubject` mess: https://github.com/macoscope/SwiftyStateMachine/commit/8540911a9e9f7633ac8902fdea7f1d2789e133f0.

I like your version better.

**Comment by bubski** [comment]
btw, just pointing out that with #20 we don't need `[unowned self]`:

``` swift
let schema = // ...
class MyViewController: UIViewController {
    lazy var stateMachine = { StateMachine(schema: schema, subject: self) }()
}
```

**Comment by narfdotpl** [comment]
Yes, I'm working on a pull request when I will also update the documentation. :+1:

**Comment by narfdotpl** [comment]
@bartekchlebek, I cherry-picked your fix to #24, could you take a look? :eyes:
